### PR TITLE
SONARJS-761 Declare the plugin as compliant with SonarLint

### PR DESCRIPTION
**Pull request by ivandalbosco**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -5,6 +5,11 @@
   <parent>
     <groupId>org.sonarsource.parent</groupId>
     <artifactId>parent</artifactId>
+    <!--
+     When updating to parent 37 or above, check if <version.sonar-packaging.plugin> defined in the new
```
Comment (vilchik-elena) [inline pom.xml]: @ivandalbosco We should remove this confusing comment

File: pom.xml [hunk 2]
```diff
@@ -66,6 +71,10 @@
   <properties>
     <gitRepositoryName>sonar-javascript</gitRepositoryName>
     <license.title>SonarQube JavaScript Plugin</license.title>
+    <!-- depends on API 6.0 but backward compatible with LTS -->
+    <sonarQubeMinVersion>5.6</sonarQubeMinVersion>
+    <!-- Temporary override of plugin version 1.16 inherited from parent 36 -->
```
Comment (vilchik-elena) [inline pom.xml]: @ivandalbosco and this one

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java [hunk 3]
```diff
@@ -60,19 +62,26 @@
...
       JavaScriptRulesDefinition.class,
-      JavaScriptProfile.class,
+      JavaScriptProfile.class);
 
-      UTCoverageSensor.class,
-      ITCoverageSensor.class,
-      OverallCoverageSensor.class,
+    // Do not waste resources adding coverage sensors to SonarLint, as they are handled as no-ops.
+    // Note: method Context.getRuntime is not available on 5.6 SQ server
+    if (!context.getSonarQubeVersion().isGreaterThanOrEqual(V6_0) || context.getRuntime().getProduct() != SonarProduct.SONARLINT) {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java]: @ivandalbosco do we really need to add these extensions for SQ 5.6 and product==Sonarlint?

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java [hunk 4]
```diff
@@ -123,12 +126,19 @@ protected void analyseFiles(SensorContext context, List<TreeVisitor> treeVisitor
...
+        if (context.getSonarQubeVersion().isGreaterThanOrEqual(V6_0) && context.isCancelled()) {
+          throw new CancellationException("Analysis interrupted because the SensorContext is in cancelled state");
+        }
         if (!isExcluded(inputFile.file())) {
           analyse(context, inputFile, treeVisitors);
         }
         progressReport.nextFile();
       }
       success = true;
+    } catch (CancellationException e) {
```
Comment (SonarTech) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java]: MAJOR Either log or rethrow this exception. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java (hunk 3 above)
Comment (ivandalbosco) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java]: @vilchik-elena We don't, but the distinction SQ server vs SonarLint is new. It cannot be made with the 5.6 API, context.getRuntime().getPoduct() is new to API 6.0

File: pom.xml (hunk 1 above)
Comment (ivandalbosco) [inline pom.xml]: @elena I disagree. We don't want to get stuck to sonar-packaging.plugin 1.17 forever. As soon as the parent uses 1.17, we want to stick to it and automatically get further versions 1.18. 1.19. etc.

Comment (vilchik-elena) [inline pom.xml]: @ivandalbosco how do you know that parent 37 will update sonar-packaging.plugin's version? I don't see it in backlog: https://jira.sonarsource.com/browse/PARENTOSS/fixforversion/13356

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java [hunk 5]
```diff
@@ -187,6 +198,23 @@ private void processRecognitionException(RecognitionException e, SensorContext s
...
+    if (sensorContext.getSonarQubeVersion().isGreaterThanOrEqual(V6_0)) {
+      sensorContext.newAnalysisError()
+        .onFile(inputFile)
+        .at(inputFile.newPointer(e.getLine(), 0))
+        .message(e.getMessage())
+        .save();
+    }
+  }
+
+  private void processException(Exception e, SensorContext sensorContext, InputFile inputFile) {
```
Comment (SonarTech) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java]: MAJOR Make "processException" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptSquidSensorTest.java [hunk 6]
```diff
@@ -288,6 +298,24 @@ public void visitScript(ScriptTree tree) {
     }
   }
 
+  private final class EmptyCheck extends DoubleDispatchVisitorCheck {
+
+    @Override
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptSquidSensorTest.java]: @ivandalbosco you don't have to override these 3 methods, there is no value of it

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java [hunk 7]
```diff
@@ -157,7 +157,7 @@ private static void stopProgressReport(ProgressReport progressReport, boolean su
   }
 
   private void analyse(SensorContext sensorContext, InputFile inputFile, List<TreeVisitor> visitors) {
-    ScriptTree scriptTree;
+    ScriptTree scriptTree;
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java]: @ivandalbosco ?

**Pull request merged**